### DISCUSSION TITLE
Reading files from folder contents vs reading files from git ref

### DISCUSSION THREAD

**Discussion by spenserblack** [o2sh/onefetch#1303] [General]
Using this discussion to list the pros/cons of reading from the folder (like tokei does) vs reading from a git ref (like linguist does). Feel free to comment or edit.

## From git ref

- Read from bare repo (https://github.com/o2sh/onefetch/issues/1299)
- Read only tracked files (a user scanned a dotfiles repo and was surprised that a bunch of untracked `.go` files (probably from `~/go/`) were included) (https://github.com/o2sh/onefetch/issues/1033)

## From directory

- Read untracked files (https://github.com/o2sh/onefetch/pull/1152#discussion_r1313880761)

**Comment by o2sh** [comment]
Can we do HEAD + pending changes ?

**Reply by spenserblack** [reply]
Going to have to ask @Byron about that, who wrote the majority of the initial gix behavior. I think reading pending changes would cause the same issue for https://github.com/o2sh/onefetch/issues/1033 (basically, that user was surprised that uncommitted files *were* included). But I'm having trouble imagining how "HEAD + pending changes" would be different from reading from the directory "tokei-style" right now :thinking:

**Reply by Byron** [reply]
For a lack of clear definition of 'pending changes', I'd go with something very simple as it helps my own understanding a lot. Thus, I see 3 sources for data:

* any commit or tree (adds support for bare repos and extraction of changes over time)
* the index (i.e. all tracked files in and changes to be committed)
* the worktree (like what's happening now)

If `tokei` one day supports virtual file systems or some abstraction over the filesystem, the git-based feeding could certainly be implemented. `gengo` can probably do much of that today.

**Reply by spenserblack** [reply]
Thanks @Byron! I'm not 100% sure what the *effective* difference would be between reading from the index and reading from the worktree. Specifically, how the list of files and their contents would be different between the two. Would it be that reading from the index would implicitly use git built-in behavior (like `.gitignore`) instead of needing to *explicitly* do that e.g. via the `ignore` crate?

And, yeah, `gengo` should be able to support anything that implements the [`FileSource`](https://docs.rs/gengo/0.11.1/gengo/trait.FileSource.html) trait. It's not a particularly easy trait to implement :laughing: But it basically allows you to define a type that can return an iterator over files and a state that can be used for multi-threaded iteration, as well as optionally defining any functionality for overriding the detected file attributes. So if the `gengo` built-in file sources don't work out, we can write one that's specific for `onefetch`.

**Reply by o2sh** [reply]
Thanks for the clarifications.

What are the cons of using the worktree? Does it handle bare repositories? Because as I understood the worktree exposes pending changes + untracked files which would be equivalent to the current behavior?

**Reply by o2sh** [reply]
But If It's easier, I wouldn't mind sticking to only considering tracked files at HEAD and adopting the GitHub Linguist approach entirely.

**Reply by Byron** [reply]
I'd like that, i.e. dropping the `index` approach which I'd find quite uncommon. Either everything in the worktree, sans .gitignored files, or everything in Git, but not a mix of that by considering the staging area. The difference of the staging area and the worktree probably also isn't big enough to warrant that.

> Specifically, how the list of files and their contents would be different between the two. Would it be that reading from the index would implicitly use git built-in behavior (like `.gitignore`) instead of needing to _explicitly_ do that e.g. via the `ignore` crate?

The `index` is a tree of a commit, and staged changes. This does indeed imply that `.gitignored` files aren't in there, as they are (typically) not added.

**Reply by spenserblack** [reply]
Ah, I see, thanks!

Your response reminded me of something I'd like to raise this for consideration:
In a dotfiles repository, it's not uncommon to ignore *everything* and add files with `git add -f` (this is what I do).
```gitignore
# .gitignore
*
```

This makes it harder to accidentally stage sensitive files and cleans up the `git status` output.
So, for certain repositories, reading the `.gitignore` might ignore *too much.*

**Reply by spenserblack** [reply]
Just checking: are we good to move forward with reading from HEAD?

**Reply by o2sh** [reply]
Yes, sounds good to me
